external/scoped_trace_terminate_on_thread_migrate.patch

```diff
@@ -0,0 +1,148 @@
++// Returns a path to a directory that contains ancillary data files that
++// might be used by tests. It is implementation dependent whether or not the
++// path is terminated by the directory-separator character. The directory
++// and the files in it should be considered read-only.
+ GTEST_API_ std::string SrcDir();
+ 
+ #ifdef _MSC_VER
+@@ -2224,8 +2230,8 @@ GTEST_API_ std::string SrcDir();
+ // insufficient. The macros should be preferred when possible, as they avoid
+ // most of the complexity of calling this function.
+ //
+-// The `factory` argument is a factory callable (move-constructible) object or
+-// function pointer that creates a new instance of the Test object. It
++// The `factory` argument is a factory callable (move-constructible) object
++// or function pointer that creates a new instance of the Test object. It
+ // handles ownership to the caller. The signature of the callable is
+ // `Fixture*()`, where `Fixture` is the test fixture class for the test. All
+ // tests registered with the same `test_suite_name` must return the same
+diff --git a/googletest/src/gtest.cc b/googletest/src/gtest.cc
+index a64e887c..a600df0c 100644
+--- a/googletest/src/gtest.cc
++++ b/googletest/src/gtest.cc
+@@ -44,14 +44,17 @@
+ #include <chrono>  // NOLINT
+ #include <cmath>
+ #include <cstdint>
++#include <exception>
+ #include <initializer_list>
+ #include <iomanip>
++#include <iostream>
+ #include <iterator>
+ #include <limits>
+ #include <list>
+ #include <map>
+ #include <ostream>  // NOLINT
+ #include <sstream>
++#include <thread>
+ #include <unordered_set>
+ #include <vector>
+ 
+@@ -1134,8 +1137,8 @@ class Timer {
+ 
+   // Return time elapsed in milliseconds since the timer was created.
+   TimeInMillis Elapsed() {
+-    return std::chrono::duration_cast<std::chrono::milliseconds>(
+-               clock::now() - start_)
++    return std::chrono::duration_cast<std::chrono::milliseconds>(clock::now() -
++                                                                 start_)
+         .count();
+   }
+ 
+@@ -6172,7 +6175,7 @@ void UnitTestImpl::ListTestsMatchingFilter() {
+     }
+   }
+   fflush(stdout);
+-  #if GTEST_HAS_FILE_SYSTEM
++#if GTEST_HAS_FILE_SYSTEM
+   const std::string& output_format = UnitTestOptions::GetOutputFormat();
+   if (output_format == "xml" || output_format == "json") {
+     FILE* fileout = OpenFileForWriting(
+@@ -6838,6 +6841,13 @@ void ScopedTrace::PushTrace(const char* file, int line, std::string message) {
+ 
+ // Pops the info pushed by the c'tor.
+ ScopedTrace::~ScopedTrace() GTEST_LOCK_EXCLUDED_(&UnitTest::mutex_) {
++  if (id != std::this_thread::get_id()) {
++    std::cerr << "ScopedTrace was created and destroyed on different "
++                 "std::threads, terminating to avoid segfaults, hangs, or "
++                 "silent corruption. Are you using any pika functionality that "
++                 "may yield a task after creating the ScopedTrace?\n";
```

Comment:
The error message is here. Suggestions for how we can make this as useful/informative as possible are very welcome!